### ISSUE TITLE
Add optimized read! and and write methods

### ISSUE THREAD

**Issue by nalimilan** [JuliaStrings/StringEncodings.jl#39]
https://github.com/JuliaStrings/StringEncodings.jl/pull/38 implements an optimized `readbytes!` method. The same approach could be used to make `read!` and `write` more efficient with arrays, probably by overloading `Base.unsafe_read` and `Base.unsafe_write`.

**Comment by vtjnash** [comment]
Yes, normally IO should provide a `unsafe_read` and `unsafe_write` function, if they support optimized IO
